File: src/server/crdt/shopping/shopping_list.go
package crdt

import (
	"fmt"
	"log"
	"sdle-server/crdt/generic"
	g01 "sdle-server/proto"
)

type ShoppingList struct {
	replicaID  string
	listID     string
	name       *crdt.MVReg[string]
	items      *crdt.ORMap[string, *ShoppingItem]
	dotContext *crdt.DotContext
}

func NewShoppingList(replicaID string, listID string) *ShoppingList {
	dotContext := crdt.NewDotContext()

	name := crdt.NewMVReg[string](replicaID)
	name.SetContext(dotContext)

	createItem := func(id string) *ShoppingItem { return NewShoppingItemWithEnabled(id, "", false) }
	items := crdt.NewORMap[string](replicaID, createItem)
	items.SetContext(dotContext)

	return &ShoppingList{
		replicaID:  replicaID,
		listID:     listID,
		name:       name,
		items:      items,
		dotContext: dotContext,
	}
}



func (sl *ShoppingList) ReplicaID() string {
	return sl.replicaID
}

func (sl *ShoppingList) ListID() string {
	return sl.listID
}

func (sl *ShoppingList) Name() string {
	// Assumes a single value in the MVReg
	values := sl.name.Read()
	if len(values) > 0 {
		return values[0]
	}
	return ""
}

func (sl *ShoppingList) SetName(name string) *ShoppingList {
	delta := NewShoppingList(sl.replicaID, sl.listID)

	nameDelta := sl.name.Reset()
	nameDelta.Join(sl.name.Write(name))

	delta.name = nameDelta
	delta.SetContext(nameDelta.Context())
	return delta
}

func (sl *ShoppingList) Context() *crdt.DotContext {
	return sl.dotContext
}

func (sl *ShoppingList) SetContext(ctx *crdt.DotContext) {
	sl.dotContext = ctx
	sl.name.SetContext(ctx)
	sl.items.SetContext(ctx)
}


func (sl *ShoppingList) Items() []ShoppingItem {
	result := []ShoppingItem{}

	ids := sl.items.Keys()
	for _, id := range ids {
		item := sl.items.Get(id)
		result = append(result, *item.Clone())
	}

	return result
}

func (sl *ShoppingList) GetItem(itemID string) *ShoppingItem {
	item := sl.items.Get(itemID)
	if item.IsNull() {
		return nil
	}
	return item.Clone()
}

func (sl *ShoppingList) PutItem(itemID string, name string, quantityDiff int64, acquiredDiff int64) *ShoppingList {
	delta := NewShoppingList(sl.replicaID, sl.listID)

	itemsDelta := sl.items.Apply(itemID, func(item *ShoppingItem) *ShoppingItem {
		item.SetItemID(itemID)

		itemDelta := item.SetName(name)
		itemDelta.Join(item.IncQuantity(quantityDiff))
		itemDelta.Join(item.IncAcquired(acquiredDiff))
		itemDelta.Join(item.Enable())

		return itemDelta
	})

	delta.items = itemsDelta
	delta.SetContext(itemsDelta.Context())

	return delta
}

func (sl *ShoppingList) RemoveItem(itemID string) *ShoppingList {
	delta := NewShoppingList(sl.replicaID, sl.listID)

	// Perform soft delete
	itemsDelta := sl.items.Apply(itemID, func(item *ShoppingItem) *ShoppingItem {
		itemDelta := item.Disable()
		return itemDelta
	})

	delta.items = itemsDelta
	delta.SetContext(itemsDelta.Context())

	return delta
}

func (sl *ShoppingList) Join(other *ShoppingList) {
	originalContext := sl.dotContext.Clone()

	sl.name.Join(other.name)
	sl.dotContext.Copy(originalContext)

	sl.items.Join(other.items)
	// No need to restore context here

	sl.dotContext.Join(other.dotContext)
}

func (sl *ShoppingList) Clone() *ShoppingList {
	clone := NewShoppingList(sl.replicaID, sl.listID)

	clone.name = sl.name.Clone()
	clone.items = sl.items.Clone()
	clone.SetContext(sl.dotContext.Clone())

	return clone
}

func (sl *ShoppingList) String() string {
	return fmt.Sprintf("ShoppingList{replicaID: %s, listID: %s, name: %s, dotContext: %v, items: %v}",
		sl.replicaID, sl.listID, sl.name, sl.dotContext, sl.items)
}

func (sl *ShoppingList) ToProto() *g01.ShoppingList {
	protoName := (*crdt.StringMVReg)(sl.name).ToProto()

	protoItems := make(map[string]*g01.ShoppingItem)
	ids := sl.items.Keys()

	for _, id := range ids {
		protoItems[id] = sl.items.Get(id).ToProto()
	}

	return &g01.ShoppingList{
		Id:         sl.listID,
		Name:       protoName,
		Items:      protoItems,
		DotContext: sl.dotContext.ToProto(),
	}
}

func ShoppingListFromProto(protoList *g01.ShoppingList, replicaId string) *ShoppingList {
	ctx := crdt.DotContextFromProto(protoList.GetDotContext())

	itemMap := make(map[string]*ShoppingItem)
	for id, protoItem := range protoList.GetItems() {
		itemMap[id] = ShoppingItemFromProto(protoItem, replicaId, id, ctx)
		log.Print(itemMap[id])
	}

	createItem := func(id string) *ShoppingItem { return NewShoppingItem(id, "") }
	items := crdt.NewORMapFrom(replicaId, createItem, ctx, itemMap)

	return &ShoppingList{
		replicaID:  replicaId,
		listID:     protoList.GetId(),
		name:       (*crdt.MVReg[string])(crdt.StringMVRegFromProto(protoList.GetName(), replicaId, ctx)),
		dotContext: ctx,
		items:      items,
	}
}

File: src/client/types/crdt/shopping/shopping-list.ts
import { ClientRequest, StringMVReg } from "@/lib/proto/client";
import DotContext from "../generic/dot-context";
import ORMap from "../generic/ormap";
import ShoppingItem from "./shopping-item";
import {
    ShoppingList as ShoppingListProto,
    ShoppingItem as ShoppingItemProto,
    DotContext as DotContextProto,
} from "@/lib/proto/global";
import ProtocolRequest from "@/lib/protocol/protocol-entity";
import MVReg from "../generic/mvreg";

export default class ShoppingList implements ProtocolRequest {
    private replicaId: string;
    private listId: string;
    private name: MVReg<string>;
    private items: ORMap<string, ShoppingItem>
    private dotContext: DotContext;

    constructor(replicaId: string, listId: string) {
        this.replicaId = replicaId;
        this.listId = listId;
        this.dotContext = new DotContext();

        this.name = new MVReg<string>(replicaId);
        this.name.setContext(this.dotContext);

        this.items = new ORMap<string, ShoppingItem>(
            this.replicaId,
            (replicaId: string) => new ShoppingItem(replicaId, ""),
            this.dotContext
        );
    }

    public getReplicaId(): string {
        return this.replicaId;
    }

    public getListId(): string {
        return this.listId;
    }

    public getName(): string {
        // Assume single value for name
        return this.name.read().at(0) || "";
    }

    public setName(name: string): ShoppingList {
        const delta = new ShoppingList(this.replicaId, this.listId);

        delta.name = this.name.write(name);
        delta.setContext(delta.name.getContext());

        return delta;
    }

    public getContext(): DotContext {
        return this.dotContext;
    }

    public setContext(context: DotContext): void {
        this.dotContext = context;
        this.name.setContext(context);
        this.items.setContext(context);
    }

    public getItems(): ShoppingItem[] {
        return Array.from(this.items.keys())
            .map((itemId: string) => this.items.get(itemId).clone())
            .filter((item: ShoppingItem) => !item.isDeleted());
    }

    public getItem(itemId: string): ShoppingItem | undefined {
        const item = this.items.get(itemId);
        if (item.isNull()) {
            return undefined;
        }
        return item.clone();
    }

    public putItem(itemId: string, quantityDiff: number, acquiredDiff: number, name?: string): ShoppingList {
        const delta = new ShoppingList(this.replicaId, this.listId);

        const itemsDelta = this.items.apply(itemId, (item: ShoppingItem) => {
            item.setItemId(itemId);

            const itemDelta = item.incQuantity(quantityDiff);
            itemDelta.join(item.incAcquired(acquiredDiff));
            itemDelta.join(item.restore());

            if (name !== undefined) {
                itemDelta.join(item.setName(name));
            }

            return itemDelta;
        });

        delta.items = itemsDelta;
        delta.setContext(itemsDelta.getContext());

        return delta;
    }

    public removeItem(itemId: string): ShoppingList {
        const delta = new ShoppingList(this.replicaId, this.listId);

        const itemsDelta = this.items.apply(itemId, (item: ShoppingItem) => {
            return item.delete();
        });

        delta.items = itemsDelta;
        delta.setContext(itemsDelta.getContext());

        return delta;
    }

    public join(other: ShoppingList): void {
        const originalCtx = this.dotContext.clone();

        this.name.join(other.name);
        this.dotContext.copy(originalCtx);

        this.items.join(other.items);
        // No need to restore context here

        this.dotContext.join(other.getContext());
    }

    public clone(): ShoppingList {
        const clone = new ShoppingList(this.replicaId, this.listId);

        clone.name = this.name.clone();
        clone.items = this.items.clone();
        clone.setContext(this.dotContext.clone());

        return clone;
    }

    public toProto(): ShoppingListProto {
        const nameProto = this.name.toStringProto();

        const itemProtos: { [key: string]: ShoppingItemProto } = {};

        for (const itemId of this.items.keys()) {
            const item = this.items.get(itemId);
            itemProtos[itemId] = item.toProto();
        }

        return ShoppingListProto.create({
            id: this.listId,
            name: nameProto,
            items: itemProtos,
            dotContext: this.dotContext.toProto(),
        });
    }

    public static fromProto(proto: ShoppingListProto, replicaId: string): ShoppingList {
        const ctx = DotContext.fromProto(proto.dotContext as DotContextProto);

        const itemMap = new Map<string, ShoppingItem>();
        for (const itemId in proto.items) {
            const itemProto = proto.items[itemId] as ShoppingItemProto;
            const item = ShoppingItem.fromProto(itemProto, replicaId, itemId, ctx);
            itemMap.set(itemId, item);
        }

        const items = new ORMap<string, ShoppingItem>(
            replicaId,
            (replicaId: string) => new ShoppingItem(replicaId, ""),
            ctx,
            itemMap
        );

        const shoppingList = new ShoppingList(replicaId, proto.id);
        shoppingList.setContext(ctx);

        shoppingList.name = MVReg.fromStringProto(proto.name as StringMVReg, replicaId, ctx);
        shoppingList.name.setContext(ctx);

        shoppingList.items = items;
        shoppingList.items.setContext(ctx);

        return shoppingList;
    }

    public toClientRequest(): ClientRequest {
        return ClientRequest.create({
            shoppingList: this.toProto(),
        });
    }
}